### PR TITLE
misc: Where possible allow Firefox and Chrome to atomically allocate ports (#25498)

### LINKED ISSUE #25498: Add environment variables for firefox port assignments
### What would you like?

Currently it is possible to externally supply a port number to be used for remote debugging via `CYPRESS_REMOTE_DEBUGGING_PORT`:
https://github.com/cypress-io/cypress/blob/991c53244718b9ec3fe8ed3e51df292465f3a4bc/packages/server/lib/browsers/protocol.ts#L35-L39

I would like to do the same for the two ports (`foxdriver`, `marionette`) assigned here:
https://github.com/cypress-io/cypress/blob/991c53244718b9ec3fe8ed3e51df292465f3a4bc/packages/server/lib/browsers/firefox.ts#L458-L464

### Why is this needed?

`getPort` is not race free as the returned port can be bound to by other processes before the process that the port was meant for by Cypress gets around to binding to it.

If I were able to set these ports via environment variables I could externally manage which of my parallel processes gets to use which (non ephemeral) port at what time.

Ultimately I would prefer if Firefox would atomically bind to ephemeral ports and for Cypress to read those back and use them (I think at least for marionette this is supported and e.g. done by geckodriver) but that is presumably more complicated and would be worth a distinct issue.

### Other

_No response_

### PR DESCRIPTION
**Pull request by ngladitz**
Firefox and Chrome let the system allocate a port when started with `--remote-debugging-port=0` and announce the port on stderr afterwards where it is now parsed from.
    
Do something similar for Firefox's Marionette port and parse it from a message that gets written to stdout.
    
This doesn't seem possible with Firefox's `devtools.debugger.remote-port` so at least allow it to be overridden via preferences so users can implement their own race free port allocation mechanism when / where required:
```ts
            setupNodeEvents: (on, config) => {
                on('before:browser:launch', (browser, opts) => {
                    opts.preferences['devtools.debugger.remote-port'] = 12345
                    return opts;
                })
            }
```

- Closes #25498

### User facing changelog
- Allow users to override Firefox's `devtools.debugger.remote-port` via a `before:browser:launch` hook in `setupNodeEvents`.
This allows to work around issues which may result from automatically assigned / random ports.

### Steps to test

To test custom 'devtools.debugger.remote-port' override use the snippet from the description and verify that the given port is used.

For the remote debugging port in Chrome / Firefox verify port 0 is requested from the command line (`--remote-debugging-port=0`), verify a random port is used and that the connection can still be established. For Firefox's marionette port a 0 should be written to the `marionette.port` preference resulting in a random port as well.

### PR Tasks
- [x] Have tests been added/updated?
- [na] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

**Review by emilyrohrbough — CHANGES_REQUESTED** [review]
@ngladitz Thank you for opening a PR. This issue still needs to be reviewed by our team on if this is something we want to add. I'll give you an update once we talk about it.

**Review by emilyrohrbough — CHANGES_REQUESTED** [review]
@ngladitz Sorry for the delayed response...

We have a few hesitations with this change. It appears that adding the ENVs themselves on the surface seem fine to add but we don't have error handling in place if the passed port was already "busy". So this really wouldn't really fix the root issues and would cause more confusion on how to fix it. I rand a few scenarios with ports-in-use and the errors aren't super helpful. So we don't want to add these ENVs at this time.

The ideal solution would be to allow the browser to bind to a random port and we pull this dynamic port from the browser. You are right that we can do this for the marionette port. I was able to draft this up on this branch: https://github.com/cypress-io/cypress/compare/emily/ff-dynamic-marionette-port

I am not sure about the `devtools.debugger.remote-port` at this time...though looking at where we determine this port value, I see we are determining this port so early to give the user this value in the `before:browser:launch` event, which is likely why there is a race condition on when it's "free" to when we try to use it.

These port values can't be managed today without the ENVS because we aren't honoring and pulling these values back out from the user defined options. If we can pull these out from the [`launchOptions](https://github.com/cypress-io/cypress/compare/emily/ff-dynamic-marionette-port#diff-6ec303ebac972615597a2bbcc7cd2986c44cea3ff5b0eccb0b8724a8527d1972R468) we can pass them into [`firefox.setup()`](https://github.com/cypress-io/cypress/compare/emily/ff-dynamic-marionette-port#diff-6ec303ebac972615597a2bbcc7cd2986c44cea3ff5b0eccb0b8724a8527d1972R566), which would override the dynamic port allocation we are doing.

This would allow for:
```js
    setupNodeEvents (_on, config) {
      _on('before:browser:launch', (browser, opts) => {
        if (process.env.CYPRESS_FOXDRIVER_PORT) {
          opts.preferences['devtools.debugger.remote-port'] = process.env.CYPRESS_FOXDRIVER_PORT
        }

        if (process.env.CYPRESS_MARIONETTE_PORT) {
          opts.preferences['marionette.port'] = process.env.CYPRESS_MARIONETTE_PORT
        }

        return opts
      })
  },

```

**Comment by ngladitz** [comment]
@emilyrohrbough Thank you for looking into this.

> The ideal solution would be to allow the browser to bind to a random port and we pull this dynamic port from the browser. You are right that we can do this for the marionette port. I was able to draft this up on this branch: https://github.com/cypress-io/cypress/compare/emily/ff-dynamic-marionette-port

Thank you that looks good and seems to work for me as well.
I experimented with something like this but was missing `user_pref("browser.dom.window.dump.enabled", true);` in my very reduced (just Firefox without Cypress; Cypress already sets the pref) testcase. Without it the marionette port does not get output to the console.

I think this could also be done for the port that is currently covered by the environment variable `CYPRESS_REMOTE_DEBUGGING_PORT` and which is passed to Firefox via the command line option `--remote-debugging-port <port>`.
When I run Firefox with `--remote-debugging-port 0` it binds to a random port (in the example output below this happens to be 32819) and outputs it as part of the following lines:
```
WebDriver BiDi listening on ws://127.0.0.1:32819
DevTools listening on ws://127.0.0.1:32819/devtools/browser/69d2d46c-4d87-4505-b51b-cf53fa299317
```

> I am not sure about the `devtools.debugger.remote-port` at this time...though looking at where we determine this port value, I see we are determining this port so early to give the user this value in the `before:browser:launch` event, which is likely why there is a race condition on when it's "free" to when we try to use it.

With an explicit port e.g. 8081 I get the following output from Firefox:
```
Started devtools server on 8081
```

With `devtools.debugger.remote-port` set to `0` I get no corresponding output from Firefox; implying the devtools server does not get started at all.

So at least for this last case some way to manually supply the port would still be appreciated.

> These port values can't be managed today without the ENVS because we aren't honoring and pulling these values back out from the user defined options. If we can pull these out from the [`launchOptions](https://github.com/cypress-io/cypress/compare/emily/ff-dynamic-marionette-port#diff-6ec303ebac972615597a2bbcc7cd2986c44cea3ff5b0eccb0b8724a8527d1972R468) we can pass them into [`firefox.setup()`](https://github.com/cypress-io/cypress/compare/emily/ff-dynamic-marionette-port#diff-6ec303ebac972615597a2bbcc7cd2986c44cea3ff5b0eccb0b8724a8527d1972R566), which would override the dynamic port allocation we are doing.
> 
> This would allow for:
> 
> ```js
>     setupNodeEvents (_on, config) {
>       _on('before:browser:launch', (browser, opts) => {
>         if (process.env.CYPRESS_FOXDRIVER_PORT) {
>           opts.preferences['devtools.debugger.remote-port'] = process.env.CYPRESS_FOXDRIVER_PORT
>         }
> 
>         if (process.env.CYPRESS_MARIONETTE_PORT) {
>           opts.preferences['marionette.port'] = process.env.CYPRESS_MARIONETTE_PORT
>         }
> 
>         return opts
>       })
>   },
> ```

I am not sure I fully understand this last part (I think the link may have been miscopied?) but if I understand the intention of the example this does look promising. Being able to override `devtools.debugger.remote-port` would suffice for me if the other two ports could be dynamically allocated by Firefox itself.

How would you like to proceed?

**Comment by emilyrohrbough** [comment]
@ngladitz there is a lot to comprehend here - I am going to comment in the code to break this conversation up to make this a bit easier to talk back and forth on.

File: packages/server/lib/browsers/firefox.ts [hunk 1]
```diff
@@ -458,7 +458,10 @@ export async function open (browser: Browser, url: string, options: BrowserLaunc
   const [
     foxdriverPort,
     marionettePort,
-  ] = await Promise.all([getPort(), getPort()])
+  ] = await Promise.all([
+    Number(process.env.CYPRESS_FOXDRIVER_PORT) || getPort(),
+    Number(process.env.CYPRESS_MARIONETTE_PORT) || getPort(),
```
Comment (emilyrohrbough) [inline packages/server/lib/browsers/firefox.ts]: Allow Dynamic Firefox Port:

> >The ideal solution would be to allow the browser to bind to a random port and we pull this dynamic port from the browser. You are right that we can do this for the marionette port. I was able to draft this up on this branch: [emily (compare)](https://github.com/cypress-io/cypress/compare/emily/ff-dynamic-marionette-port)

> Thank you that looks good and seems to work for me as well.

Let's move forward with the draft changes I wrote up for allowing firefox to handle the dynamic port allocation.  Sounds like this works for you 👍🏻

File: packages/server/lib/browsers/firefox.ts [hunk 2]
```diff
@@ -458,7 +458,10 @@ export async function open (browser: Browser, url: string, options: BrowserLaunc
...
+  ])
```
Comment (emilyrohrbough) [inline packages/server/lib/browsers/firefox.ts]: > I think this could also be done for the port that is currently covered by the environment variable CYPRESS_REMOTE_DEBUGGING_PORT and which is passed to Firefox via the command line option --remote-debugging-port <port>.
When I run Firefox with --remote-debugging-port 0 it binds to a random port (in the example output below this happens to be 32819) and outputs it as part of the following lines:
>
> WebDriver BiDi listening on ws://127.0.0.1:32819
DevTools listening on ws://127.0.0.1:32819/devtools/browser/69d2d46c-4d87-4505-b51b-cf53fa299317

This sounds reasonable to me to allow this port to be dynamically set for the remote-bugging-port as well. We would want do do something like the following to continue to honor `CYPRESS_REMOTE_DEBUGGING_PORT`.
```
remotePort =  Number(process.env.CYPRESS_REMOTE_DEBUGGING_PORT) || '0'
```

Also we would need to confirm this works/the output is the same with Firefox 86... That is or minimum supported Firefox version.

Comment (emilyrohrbough) [reply packages/server/lib/browsers/firefox.ts]: Though are you pulling this from `stderr`? that is where I am seeing this which is odd

see: [`emily/ff-dynamic-marionette-port`#diff-6ec303ebac](https://github.com/cypress-io/cypress/compare/emily/ff-dynamic-marionette-port#diff-6ec303ebac972615597a2bbcc7cd2986c44cea3ff5b0eccb0b8724a8527d1972R568)

File: packages/server/lib/browsers/firefox.ts [hunk 3]
```diff
@@ -458,7 +458,10 @@ export async function open (browser: Browser, url: string, options: BrowserLaunc
   const [
     foxdriverPort,
     marionettePort,
-  ] = await Promise.all([getPort(), getPort()])
+  ] = await Promise.all([
+    Number(process.env.CYPRESS_FOXDRIVER_PORT) || getPort(),
```
Comment (emilyrohrbough) [inline packages/server/lib/browsers/firefox.ts]: >> These port values can't be managed today without the ENVS because we aren't honoring and pulling these values back out from the user defined options. If we can pull these out from the [launchOptions](](https://github.com/cypress-io/cypress/compare/emily/ff-dynamic-marionette-port#diff-6ec303ebac972615597a2bbcc7cd2986c44cea3ff5b0eccb0b8724a8527d1972R566)https://github.com/cypress-io/cypress/compare/emily/ff-dynamic-marionette-port#diff-6ec303ebac972615597a2bbcc7cd2986c44cea3ff5b0eccb0b8724a8527d1972R468) we can pass them into [firefox.setup()`, which would override the dynamic port allocation we are doing.

>> This would allow for:
>> ```js
>>     setupNodeEvents (_on, config) {
>>       _on('before:browser:launch', (browser, opts) => {
>>         if (process.env.CYPRESS_FOXDRIVER_PORT) {
>>           opts.preferences['devtools.debugger.remote-port'] = process.env.CYPRESS_FOXDRIVER_PORT
>>         }
>> 
>>         if (process.env.CYPRESS_MARIONETTE_PORT) {
>>           opts.preferences['marionette.port'] = process.env.CYPRESS_MARIONETTE_PORT
>>         }
>> 
>>         return opts
>>       })
>>   },
>> ```

> I am not sure I fully understand this last part (I think the link may have been miscopied?) but if I understand the intention of the example this does look promising. Being able to override devtools.debugger.remote-port would suffice for me if the other two ports could be dynamically allocated by Firefox itself.

Sorry for this confusing message -- rereading it took me a minute to figure out what the heck I was saying....

Comment (emilyrohrbough) [reply packages/server/lib/browsers/firefox.ts]: I dug into the code a big more, we will need to allocation a random port to provide to the launch options to successfully start the devtools server (see: the [`-start-debugger-server` comment](https://github.com/cypress-io/cypress/blob/3c28116d49fb6684e6306ced2738ea9750112206/packages/server/lib/browsers/firefox.ts#L409)).

The default port Firefox uses in 6000, per `/Applications/Firefox.app/Contents/MacOS/firefox --help` on my Mac, but we are overriding this value with `devtools.debugger.remote-port`
>   --start-debugger-server [ws:][ <port> | <path> ] Start the devtools server on
                     a TCP port or Unix domain socket path. Defaults to TCP port
                     6000. Use WebSocket protocol if ws: prefix is specified.

What I was trying to say above, is we should update the logic in `/packages/server/lib/browsers/firefox.ts` to use the user-defined `opts.preferences['devtools.debugger.remote-port'] ` and `opts.preferences['marionette.port']` values if defined by the user. This would remove the need for additional ENVs to allow a user to set these themselves.

The beforeLaunch values are returned as `launchOptions` here: https://github.com/cypress-io/cypress/blob/3c28116d49fb6684e6306ced2738ea9750112206/packages/server/lib/browsers/firefox.ts#L468

However, we aren't honoring these updated values when we are connecting to the launched firefox browser instance here: https://github.com/cypress-io/cypress/blob/3c28116d49fb6684e6306ced2738ea9750112206/packages/server/lib/browsers/firefox.ts#L566

Comment (emilyrohrbough) [reply packages/server/lib/browsers/firefox.ts]: see: [`emily/ff-dynamic-marionette-port`#diff-6ec303ebac](https://github.com/cypress-io/cypress/compare/emily/ff-dynamic-marionette-port#diff-6ec303ebac972615597a2bbcc7cd2986c44cea3ff5b0eccb0b8724a8527d1972R586)

File: packages/server/lib/browsers/firefox.ts (hunk 2 above)
Comment (ngladitz) [reply packages/server/lib/browsers/firefox.ts]: I was looking at both streams combined and didn't realize the messages were output to distinct streams but now that I am looking for it that does seem to be the case. It does feel a little inconsistent but I also see this in Firefox code:

```Cu.printStderr(`WebDriver BiDi listening on ${this.address}\n`);```
and
```Cu.printStderr(`DevTools listening on ${this.address}\n`);```
but for Marionette
```lazy.logger.info(`Listening on port ${this.port}`);```
which with `browser.dom.window.dump.enabled` apparently outputs to stdout.

I tried looking at Firefox 86 code but I am not that familiar with Mercurial.
What I think I see is that `WebDriver BiDi listening on ` does not exist yet but the `DevTools listening on` message which also contains the same port could work. The Marionette port message looks like it could work (same string / message but different output code).

Comment (emilyrohrbough) [reply packages/server/lib/browsers/firefox.ts]: Interesting.. Wonder why it's printing to stderr?? Maybe because it's meant for debugging and not "production" browser launches 😅

Comment (ngladitz) [reply packages/server/lib/browsers/firefox.ts]: FWIW I did find it (the fact that it outputs to stderr, the message and I suppose intended use) documented here:
https://firefox-source-docs.mozilla.org/remote/cdp/Usage.html
Also seem to say that this happens to be implemented since Firefox 86 exactly :smiley:

Comment (emilyrohrbough) [reply packages/server/lib/browsers/firefox.ts]: these docs are always hard to find! nice!

**Comment by emilyrohrbough** [comment]
@ngladitz Just curious, do you have any updates with this PR?

**Comment by ngladitz** [comment]
> @ngladitz Just curious, do you have any updates with this PR?

I hope I didn't misunderstand but the code changes in my branch associated with this PR are entirely obsoleted by the more sensible approach you implemented in your own `ff-dynamic-marionette-port` branch. I think the only thing missing is parsing the remote debugging port from the `DevTools listening on` message rather than the `WebDriver BiDi listening on` message(?).

**Comment by emilyrohrbough** [comment]
@ngladitz I was hoping you'd be able to borrow some of these change and pull them in to your change set. These changes I played with won't be something I can prioritize working on fully to get across the line. Sorry for the confusion.

**Comment by ngladitz** [comment]
@emilyrohrbough sorry about that. Based on your branch I generalized the changes to also cover Chrome's remote debugging port (apparently even the message with the port output by both browsers is the same).

I tested everything locally through manual testing as far as I could follow the possible implications but where I am unsure is Chrome with an existing instance. I hope this is covered by automated testing / CI as I don't really know how it is used exactly.

Also not really sure how automated testing and CI work here; I see failures but I can't tell what they are or if they are related to my changes.

**Comment by mschile** [comment]
@ngladitz, looks like some of the test failures ([server-unit-tests](https://app.circleci.com/pipelines/github/cypress-io/cypress/51781/workflows/2afd2fea-e157-4fee-ada2-3568153d6d4b/jobs/2136615?invite=true#step-111-2187) and [server-integration-tests](https://app.circleci.com/pipelines/github/cypress-io/cypress/51781/workflows/2afd2fea-e157-4fee-ada2-3568153d6d4b/jobs/2136634?invite=true#step-111-5857)) are from this PR along with some [linting errors](https://app.circleci.com/pipelines/github/cypress-io/cypress/51781/workflows/2afd2fea-e157-4fee-ada2-3568153d6d4b/jobs/2136612?invite=true#step-111-157).

To run the server unit-tests:
`yarn workspace @packages/server test-unit chrome_spec`
`yarn workspace @packages/server test-unit firefox_spec`

To run the server integration tests:
`yarn workspace @packages/server test-integration cypress_spec`

**Comment by ngladitz** [comment]
@mschile Thank you for the pointers. I Think I managed to fix the tests you indicated (at least locally).
Mostly stubs not implementing the new browser requirements.
I also dropped some pre-86 Firefox assumptions which I assume should not be relevant since >= 86 is required.

I put the changes into a new commit as I wasn't sure of what you'd prefer. I've been rebasing / rewriting history earlier but given that 'develop' was merged back into my branch I guess you might prefer a new commit?

**Comment by mschile** [comment]
> I put the changes into a new commit as I wasn't sure of what you'd prefer. I've been rebasing / rewriting history earlier but given that 'develop' was merged back into my branch I guess you might prefer a new commit?

Yes, in general a new commit is preferred. We'll squash merge the PR into develop at the end.

File: packages/server/lib/browsers/chrome.ts [hunk 4]
```diff
@@ -700,10 +700,15 @@ export = {
     // first allows us to connect the remote interface,
     // start video recording and then
     // we will load the actual page
-    const launchedBrowser = await launch(browser, 'about:blank', port, args, launchOptions.env) as unknown as BrowserInstance & { browserCriClient: BrowserCriClient}
+    const process = launch(browser, 'about:blank', port, args, launchOptions.env)
```
Comment (AtofStryker) [inline packages/server/lib/browsers/chrome.ts]: doesn't this still need to be awaited? Why the change here?

**Review by AtofStryker — COMMENTED** [review]
Giving CI another run because the failures look related to a missing `.node_version` file, which should be unrelated here.

File: packages/server/lib/browsers/chrome.ts (hunk 4 above)
Comment (ngladitz) [reply packages/server/lib/browsers/chrome.ts]: > doesn't this still need to be awaited? Why the change here?

The launch function doesn't return a promise. There was a stub that was async (which I changed to non-async) but the actual implementation wasn't. The Firefox implementation called launch without await so I dropped it in the Chrome implementation as well.

**Comment by mschile** [comment]
Hi @ngladitz, would you be able to update the PR description with the most up to date information.

**Comment by ngladitz** [comment]
@mschile I've updated the description and fixed the `check-ts` test (I dropped the unused port argument from the `launch` signature but apparently forgot to drop it from one of the call sites). I managed to run the test locally which now seems to pass but I am unsure how to interpret or run (locally or otherwise) the rest of the failing tests.

**Comment by cacieprins** [comment]
Hi @ngladitz ! Thank you for updating the description, and fixing the `check-ts` test. Have you made any progress on resolving the failing unit tests? You can run these tests on your forked branch with the following commands:

For server unit tests:
`yarn workspace @packages/server test-unit chrome_spec`
`yarn workspace @packages/server test-unit firefox_spec`

To run the server integration tests:
`yarn workspace @packages/server test-integration cypress_spec`

**Comment by ngladitz** [comment]
@cacieprins thank you. @mschile did mention those before and I was so sure I had them passing locally but rerunning them now they certainly don't. I must have missed a step somewhere.

I pushed a new commit and they all pass now except for the following integration test:
```
  1) lib/cypress --run-project logs error and exits when project folder has read permissions only and cannot write cypress.config.js:

      error was logged
      + expected - actual

       [
      -  "Error: Can't run because \u001b[95mno spec files\u001b[39m were found.\n\nWe searched for specs matching this glob pattern:\n\n\u001b[90m  > \u001b[39m\u001b[94m/home/ngladitz/src/git/cypress/packages/server/permissions/\u001b[33mcypress/e2e/**/*.cy.{js,jsx,ts,tsx}\u001b[94m\u001b[39m"
      +  {
      +    "type": "ERROR_WRITING_FILE"
      +  }
       ]
```

The description of the test seems to suggest that the expectation is that removing write permissions to the directory containing a `cypress.config.js` would remove permissions to the file itself as well ... which it doesn't so the test failure seems justified as far as I can tell. I might misinterpret the description / intention (doesn't cypress just consume the file? why does it need write permissions?) or maybe this is something system specific. Either way I presume this failure has nothing to do with my changes given that I haven't touched anything that would feel related(?)
